File: Daily Resources/Day03/DSA - Dynamic Programming Solutions/partition_equal_sum.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (int x : nums) sum += x;
        if (sum % 2) return false;
        int target = sum / 2;
        vector<bool> dp(target + 1);
        dp[0] = true;
        for (int x : nums) {
            for (int t = target; t >= x; --t) {
                dp[t] = dp[t] || dp[t - x];
            }
        }
        return dp[target];
    }
};

File: Daily Resources/Day03/DSA - Dynamic Programming Solutions/partition_equal_sum.py
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        s = sum(nums)
        if s % 2:
            return False
        target = s // 2
        dp = [True] + [False] * target
        for x in nums:
            for t in range(target, x - 1, -1):
                dp[t] = dp[t] or dp[t - x]
        return dp[target]
